### PR TITLE
[Reset] Handle reset run in processStartChildExecution

### PR DESCRIPTION
**Pull request by gow**
## What changed?
Updated processStartChildExecution() to handle reset runs. In this change we describe the child workflow first to verify parent-child relationship before reconnecting to it. 

## Why?
For now, we default to 'reconnect' option if there is a pending child at the time of reset. But first we need to verify that the child is indeed related to the current run before reconnecting to it. The verification is done by comparing `OriginalRunID` values in the `ExecutionInfo`

## How did you test it?
Added unit test
The test doesn't cover all possible scenarios yet. It was hard to mock the child's parent's mutable state since there is no direct handle to it. So I'll cover that path in a feature test when I get to the end of the feature development.
However, I hard coded some errors and reset a workflow to manually verify those cases.

## Potential risks
N/A 

## Documentation
N/A

## Is hotfix candidate?
No

### CODE REVIEW

File: service/history/transfer_queue_active_task_executor.go [hunk 1]
```diff
@@ -838,6 +839,40 @@ func (t *transferQueueActiveTaskExecutor) processStartChildExecution(
 		}
 	}
 
+	// Note: childStarted flag above is computed from the parent's history. When this is TRUE it's guaranteed that the child was succesfully started.
+	// But if it's FALSE then the child *may or maynot* be started (ex: we failed to record ChildExecutionStarted event previously.)
+	// This is not an issue in the general flow since starting a child is an idempotent operation (guarded by childInfo.CreateRequestId)
+	// But in the case of resets, we overwrite childInfo.CreateRequestId with a newly generated ID. Hence we need special handling when we are
```
Comment (yycptt) [inline service/history/transfer_queue_active_task_executor.go:845]: I think this doesn't matter as the createRequestID check when starting a workflow only applies the latest run. 

To really solve the issue, StartWorkflow api needs to support start a specific runID and return an error if that run already exists.

File: service/history/workflow/mutable_state_impl.go [hunk 2]
```diff
@@ -776,6 +776,12 @@ func (ms *MutableStateImpl) UpdateResetRunID(runID string) {
 	ms.executionInfo.ResetRunId = runID
 }
 
+// IsResetRun returns true if this run is the result of a reset operation.
+// A run is a reset run if OriginalExecutionRunID points to another run.
+func (ms *MutableStateImpl) IsResetRun() bool {
+	return ms.GetExecutionInfo().GetOriginalExecutionRunId() != ms.GetExecutionState().GetRunId()
```
[the remark this answers is not in the document]
Comment (gow) [reply service/history/workflow/mutable_state_impl.go:782]: Luckily no. Current RunID is passed on as OriginalRunID when we ContinueAsNew
https://github.com/temporalio/temporal/blob/c8b91e1317b08500f3cae01357f2fc936d2b3d2e/service/history/workflow/mutable_state_impl.go#L2224

File: service/history/transfer_queue_active_task_executor.go [hunk 3]
```diff
@@ -838,6 +839,40 @@ func (t *transferQueueActiveTaskExecutor) processStartChildExecution(
...
+	// in the reset run before we can attempt to start the child.
+	if mutableState.IsResetRun() {
+		childRunID, err := t.checkAndReconnectChild(ctx, mutableState, targetNamespaceEntry, attributes.WorkflowId)
+		if err != nil {
+			return err
+		}
+		if childRunID != "" {
```
[the remark this answers is not in the document]
Comment (gow) [reply service/history/transfer_queue_active_task_executor.go:850]: Done. Added it below.

File: service/history/transfer_queue_active_task_executor.go (hunk 1 above)
Comment (gow) [reply service/history/transfer_queue_active_task_executor.go]: Done. Updated the comment.

File: service/history/transfer_queue_active_task_executor.go [hunk 4]
```diff
@@ -908,6 +946,76 @@ func (t *transferQueueActiveTaskExecutor) processStartChildExecution(
 	}, parentClock, childClock)
 }
 
+// verifyChildWorkflow describes the childWorkflowID and identifies its parent. It then checks if the current run was derived from that parent by comparing the OriginalRunID value.
+// It returns the child's runID if the checks pass. Empty runID is returned if the check doesn't pass.
+func (t *transferQueueActiveTaskExecutor) verifyChildWorkflow(
+	ctx context.Context,
+	// task *tasks.ResetWorkflowTask,
```
Comment (yycptt) [inline service/history/transfer_queue_active_task_executor.go:953]: nit: remove

File: service/history/transfer_queue_active_task_executor.go [hunk 5]
```diff
@@ -838,6 +839,43 @@ func (t *transferQueueActiveTaskExecutor) processStartChildExecution(
...
+
+			parentClock, err := t.shardContext.NewVectorClock()
+			if err != nil {
+				return err
+			}
+			return t.createFirstWorkflowTask(ctx, task.TargetNamespaceID, childExecution, parentClock, childClock)
+		}
+		// now if there was no child found after reset then it could mean one of the following.
+		// 1. The parent never got a chance to start the child. So we should go ahead and start one (below)
+		// 2. The child was started, but may be terminated from someone external.
```
Comment (yycptt) [inline service/history/transfer_queue_active_task_executor.go:874]: nit: terminated or timedout

File: service/history/transfer_queue_active_task_executor.go [hunk 6]
```diff
@@ -908,6 +946,76 @@ func (t *transferQueueActiveTaskExecutor) processStartChildExecution(
...
+		return "", nil
+	}
+
+	childsParentRunID := response.WorkflowExecutionInfo.ParentExecution.RunId
+	// Check if the child's parent was the base run for the current run.
+	if childsParentRunID == mutableState.GetExecutionInfo().OriginalExecutionRunId {
+		return response.WorkflowExecutionInfo.Execution.RunId, nil
+	}
+
+	// load the child's parent mutable state.
```
Comment (yycptt) [inline service/history/transfer_queue_active_task_executor.go:991]: We could consider storing parent original runID in child's mutable state as well. but probably not necessary for this low volume case

**Review by yycptt — APPROVED** [review]

File: service/history/transfer_queue_active_task_executor.go (hunk 6 above)
Comment (gow) [reply service/history/transfer_queue_active_task_executor.go:991]: Agree.

**Pull request merged**